### DISCUSSION TITLE
PWM Heater

### DISCUSSION THREAD

**Discussion by Bannhammer** [theyosh/TerrariumPI#965] [Q&A]
Hello.
I have a problem with the dimming function. The dimmer only ever regulates from 0 to 100 percent.
Can someone maybe take a look at my settings?

![PWM](https://github.com/user-attachments/assets/f99959f8-f5a7-4156-9f08-cabd8ef6ef3f)
![set1](https://github.com/user-attachments/assets/e37d412c-0a3d-4943-872a-27593824788a)
![set2](https://github.com/user-attachments/assets/e9c063b0-3252-49a2-84a6-36ffd9b4d36a)

**Comment by theyosh** [comment]
I guess putting a `0` in the field `Wartezeit` should fix it.

Other fields looks good.

**Reply by theyosh** [reply]
Also, the dimmer will go to 100% in the beginning. So that is normal. But after an hour or so, it should level out at the average temperature.

https://en.wikipedia.org/wiki/Proportional%E2%80%93integral%E2%80%93derivative_controller#Controller_theory

**Comment by Bannhammer** [comment]
Thanks Yosh for the reply. I will try that out. I will report.

**Comment by Bannhammer** [comment]
Hi Yosh. 
I let the PWM run overnight with the settings but it doesn't settle. I have a controller from Amazon . 
https://www.amazon.de/dp/B0BLS653TW?ref=ppx_yo2ov_dt_b_fed_asin_title
Could it also be because he doesn't regulate properly?

![neu](https://github.com/user-attachments/assets/b886e528-84e1-4355-95a7-35e9d375f4c8)

**Reply by theyosh** [reply]
Not sure if this dimmer will work. I see some comments below the article that are saying it is not the best device.

On which GPIO pin have you connected this PWM dimmer? It should be either 12, 32, 33, 35 https://pinout.xyz/pinout/pwm

And the easiest way to test it, is to use a light bulb. Because than you can just turn the dimmer to 50% and see what the light does.
Other devices like heating maps, it is hard to determine if it is running at 50% as you can't see it.

**Comment by Bannhammer** [comment]
I connected the dimmer to my PCA9685 module.
Manual dimming works. I tested with a light bulb and measured with a multimeter. Even if I don't have the PCA9685 in between and go to GPIO pin 32, it dims but the process looks the same.

**Reply by theyosh** [reply]
can you post a screen shot of the sensor graph? And also the entered values?

**Reply by theyosh** [reply]
Do you mean that manual dimming works in TerrariumPI? Of did you test it differently?

**Comment by Bannhammer** [comment]
Yes, I manually dimmed Terrariumpi to 50% and the light bulb became darker and the multimeter showed me around 110V. So I think manual dimming works.
![sensor](https://github.com/user-attachments/assets/253a9e7b-60ba-4097-a8ab-35a460ebabd4)

**Comment by Bannhammer** [comment]
![senset](https://github.com/user-attachments/assets/fe04b816-f92a-4aef-8f1b-8368f7fd5b04)

**Reply by theyosh** [reply]
I would use an Alarm min value of 22 and Alarm max value of 27. So that the average temperature would get around 25,5 degrees. I think this will make the dimmer stay more at a average power level. Now the min and max are the same, what could cause a strange outcome for the dimmer power.

But based on the sensor graph, it looks like the temperature is always around 25 degrees. So that does look good despite the strange power graph.

**Comment by Bannhammer** [comment]
Ok i will give it a try. Thanks

**Comment by Bannhammer** [comment]
Somehow it doesn't quite work yet.
The goal is to get a constant temperature.
![senset](https://github.com/user-attachments/assets/112ff5aa-b177-4417-8219-21eb7033f5de)

**Reply by theyosh** [reply]
Could also post the sensor graph again?

**Comment by Bannhammer** [comment]
![sensor](https://github.com/user-attachments/assets/d3fab27d-00ef-48a7-83b3-ec93e3e3b271)

**Comment by theyosh** [comment]
I get the feeling that your heater is either a bit slow, or lacking power. Because the temperature looks pretty good. It stays withing the alarm values, and it does average at around 27 degrees.

As the relay graphs show, it does not go fully on, but tries to ramp up the power. And than it stays at 100%. Which give me the idea that the heater is a bit slow, or just not enough power. So that would cause some strange effect in the calculation for the power. And I guess that also causes the strange relay graph.

But the goal is the temperature, and that looks good. I would just ignore the relay graphs, and look at the temperature.
You could try to increase the lower alarm with 1 and degrees the high alarm with 1, so the area between the alarm min and max will be less big. You can 'play' with the alarm values in order to see what is happening. But do not enter the same value for the low and high alarm.

**Comment by Bannhammer** [comment]
I'll try a different heater.
I first attached a heater to the side window and thought that might be the problem. Then I took a heat panel. But I think that's also a bit weak. I have a stronger one and will try it out.

**Reply by theyosh** [reply]
Just curious, does a different heater work better?

**Comment by Bannhammer** [comment]
Hi Josh.
Another heater didn't bring any improvement. But I also have to say I didn't test any further after that. I'll look into the matter again this weekend. I also have a dimmable thermostat and everything works.

**Comment by Bannhammer** [comment]
Hi Josh.
I tested a little again. I connected another light bulb to the PWM. The light bulb only starts to light up at around 20 percent and then reaches 100 percent power at 30 percent.
I can imagine that this is my mistake. But how can I fix this?

**Reply by theyosh** [reply]
Hmm, which dimmer device are you using? Still https://www.amazon.de/dp/B0BLS653TW?ref=ppx_yo2ov_dt_b_fed_asin_title
I see a few review which are not that happy.

And this dimmer is not tested with me, so I can't tell how it should work. I hope it is a PWM dimmer, but I get the feeling that this is to cheap, and therefore not the best one.

One thing I can think of, is to use an Arduino in between. But this will make it more difficult. Because that means that you have to create some Arduino code that provides a webserver so TP can connect to it, and some Arduino code for the dimming part.
This way you can make your own remote dimmer, which can be used in TP.

But for now, I do not know what to do.

You could try to play with the used frequencies

![image](https://github.com/user-attachments/assets/f61c09cd-6fb5-49c1-bf17-3cad6ac9b95a)

I would make some small Python code to directly controll the dimmer for testing. But that is hard to do remotely. And I do not have example code laying around...

**Reply by Bannhammer** [reply]
Hi Josh.
I got another PWM but here too it's exactly the same that the lamp lights up at 30 percent full.

Here is the link to the dimmer.
https://www.ebay.de/itm/124296306004

**Reply by theyosh** [reply]
Just ordered one so I can do some testing.

**Reply by theyosh** [reply]
For me, this dimmer does not work. It switches on at around 13% duty. And no dimming as I can see. I emailed the seller to see if it is a software issue or not.

**Reply by Bannhammer** [reply]
Ahh okay. It seems that is the issue. Which dimmer do you use?
Maybe the seller will answer.

**Reply by theyosh** [reply]
> Ahh okay. It seems that is the issue. Which dimmer do you use? Maybe the seller will answer.

https://www.ebay.de/itm/124296306004
